core/src/main/scala/org/apache/spark/rpc/RpcEndpointRef.scala

```diff
@@ -46,6 +46,17 @@ private[spark] abstract class RpcEndpointRef(conf: SparkConf)
    */
   def send(message: Any): Unit
 
+  /**
+   * Send a message to the corresponding [[RpcEndpoint.receiveAndReply)]] and return a [[Future]] to
+   * receive the reply within the specified timeout.
+   * Return a `CancelableFuture` instance which wrap `Future` but with additional `cancel` method.
+   *
+   * This method only sends the message once and never retries.
+   */
+  def askCancelable[T: ClassTag](message: Any, timeout: RpcTimeout): CancelableFuture[T] = {
```

Comment:
Hm, I understand why we need this. But not sure whether we should add this interface. Are we really able to cancel a request that has been sent out?

Reply:
@zsxwing 
Although we cannot withdraw message sent out, I hope to do similar "cleanup" stuff like what "RPC timeout" will do. See code here
https://github.com/apache/spark/blob/2eda1876337e65915f03464076a772ea809bd361/core/src/main/scala/org/apache/spark/rpc/netty/NettyRpcEnv.scala#L248
The `rpcMessage.onCancel()` will do the same thing with `rpcMessage.onTimeout()`

If we do not add `cancel` interface, we have no way to trigger the code `rpcMessage.onCancel()` because we cannot access the `rpcMessage` instance from the `Future` instance.

Reply:
So this is not really cancelable, it just does cleanup when the request is cancelled. How about give it another name? - Like `askWithCleanupOnCancel` ?

Reply:
I am ok with the name either `askWithCleanupOnCancel` or `askCancelable`
When at client side we do the rpc cleanup "remove request id from `outstandingRpcs`", then the rpc callback will be ignored when client receive its response.
https://github.com/apache/spark/blob/b3394db1930b3c9f55438cb27bb2c584bf041f8e/common/network-common/src/main/java/org/apache/spark/network/client/TransportResponseHandler.java#L189
Not very strict, this behavior, from the caller side, it could be regarded as "canceled", because once it is cleaned, no responding callback will happen.

Reply:
@zsxwing Any thoughts ?

Reply:
@zsxwing Or we can use some other name, like `abort` instead of `cancel` ?

Reply:
`abort` sounds better